File: evals/harness/run_story_review.py
#!/usr/bin/env python3
"""End-to-end BMAD story-review eval (implementation.md §9 Phase 2).

Runs the §7.7 judge over the story_review golden set and reports:
  - first-pass rate accuracy (judge verdict vs expected_verdict)
  - failure-taxonomy accuracy (failure_kind vs label, on reworks)
  - gate-level agreement (3 nouls)

Usage:
  python3 evals/harness/run_story_review.py
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent
sys.path.insert(0, str(ROOT / "router"))
sys.path.insert(0, str(Path(__file__).resolve().parent))

from judge import GATE_THRESHOLD, judge  # noqa: E402

GOLDEN = Path(__file__).resolve().parent.parent / "golden-sets" / "story_review" / "story_review.golden.jsonl"
RESULTS = Path(__file__).resolve().parent.parent / "results"


def main():
    examples = [json.loads(l) for l in GOLDEN.read_text().splitlines() if l.strip()]
    verdict_ok = taxonomy_ok = taxonomy_n = 0
    gate_hits = gate_n = 0
    per_example = []
    for ex in examples:
        res = judge(ex["story"], ex["implementation"])
        v_ok = res["verdict"] == ex["expected_verdict"]
        verdict_ok += v_ok
        if ex["expected_verdict"] == "rework":
            taxonomy_n += 1
            taxonomy_ok += res["failure_kind"] == ex["labels"]["failure_kind"]
        for g in ("gate_spec", "gate_no_regression", "gate_security"):
            gate_n += 1
            pred_safe = res["gate_nouls"][g] >= GATE_THRESHOLD
            gate_hits += pred_safe == bool(ex["labels"][g])
        per_example.append({"id": ex["id"], "expected": ex["expected_verdict"],
                            "got": res["verdict"], "ok": v_ok,
                            "failure_kind": res["failure_kind"],
                            "gate_nouls": res["gate_nouls"],
                            "gate_labels": {g: ex["labels"][g] for g in
                                            ("gate_spec", "gate_no_regression", "gate_security")},
                            "dimensions": res["dimensions"]})

    report = {
        "set": "story_review",
        "n_examples": len(examples),
        "verdict_accuracy": verdict_ok / len(examples),
        "taxonomy_accuracy": (taxonomy_ok / taxonomy_n) if taxonomy_n else None,
        "gate_agreement": gate_hits / gate_n,
        "per_example": per_example,
    }
    RESULTS.mkdir(parents=True, exist_ok=True)
    out = RESULTS / f"story-review-{__import__('time').strftime('%Y%m%d-%H%M%S')}.json"
    out.write_text(json.dumps(report, indent=2), encoding="utf-8")
    print(json.dumps({k: v for k, v in report.items() if k != "per_example"}, indent=2))
    for p in per_example:
        mark = "ok " if p["ok"] else "MISS"
        print(f"  [{mark}] {p['id']}: expected={p['expected']} got={p['got']} kind={p['failure_kind']}")
    print(f"report: {out}")


if __name__ == "__main__":
    main()

File: evals/harness/run_story_review.ts
#!/usr/bin/env node
// End-to-end BMAD story-review eval (implementation.md §9 Phase 2).
//
// Runs the §7.7 judge over the story_review golden set and reports:
//   - first-pass rate accuracy (judge verdict vs expected_verdict)
//   - failure-taxonomy accuracy (failure_kind vs label, on reworks)
//   - gate-level agreement (3 nouls)
//
// Usage:
//   npx tsx evals/harness/run_story_review.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { GATE_THRESHOLDS, judge } from "../../router/judge.ts";
import { pyDumpsIndent, tagPythonFloats, type Json } from "./jev_policy.ts";

const HERE = path.dirname(path.resolve(import.meta.url.replace(/^file:\/\//, "")));
// PyFloat-tagged values (from judge()) unwrap to plain numbers for comparisons
const un = (v: unknown): number =>
  v !== null && typeof v === "object" && "v" in (v as Record<string, unknown>) ? ((v as Record<string, unknown>).v as number) : (v as number);
const GOLDEN = path.resolve(HERE, "..", "golden-sets", "story_review", "story_review.golden.jsonl");
const RESULTS = path.resolve(HERE, "..", "results");

function pyStamp(): string {
  const d = new Date();
  const p = (n: number, w = 2) => String(n).padStart(w, "0");
  return `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}-${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`;
}

async function main(): Promise<void> {
  const examples = fs.readFileSync(GOLDEN, "utf-8").split("\n").filter((l) => l.trim()).map((l) => JSON.parse(l));
  let verdictOk = 0, taxonomyOk = 0, taxonomyN = 0, gateHits = 0, gateN = 0;
  const perExample: any[] = [];
  for (const ex of examples) {
    const res = await judge(ex.story, ex.implementation);
    const vOk = res.verdict === ex.expected_verdict;
    verdictOk += vOk ? 1 : 0;
    if (ex.expected_verdict === "rework") {
      taxonomyN += 1;
      taxonomyOk += res.failure_kind === ex.labels.failure_kind ? 1 : 0;
    }
    for (const g of ["gate_spec", "gate_no_regression", "gate_security"]) {
      gateN += 1;
      const predSafe = un(res.gate_nouls[g]) >= un(GATE_THRESHOLDS[g]);
      gateHits += predSafe === Boolean(ex.labels[g]) ? 1 : 0;
    }
    const gn: Record<string, unknown> = {};
    for (const g of ["gate_spec", "gate_no_regression", "gate_security"]) gn[g] = un(res.gate_nouls[g]);
    const dims: Record<string, unknown> = {};
    for (const d of Object.keys(res.dimensions as Record<string, unknown>)) dims[d] = Math.trunc(un((res.dimensions as Record<string, unknown>)[d]));
    perExample.push({ id: ex.id, expected: ex.expected_verdict,
      got: res.verdict, ok: vOk,
      failure_kind: res.failure_kind,
      gate_nouls: gn,
      gate_labels: Object.fromEntries(["gate_spec", "gate_no_regression", "gate_security"].map((g) => [g, ex.labels[g]])),
      dimensions: dims });
  }

  const report: Record<string, Json> = {
    set: "story_review",
    n_examples: examples.length,
    verdict_accuracy: verdictOk / examples.length,
    taxonomy_accuracy: taxonomyN ? taxonomyOk / taxonomyN : (null as unknown as Json),
    gate_agreement: gateHits / gateN,
    per_example: perExample as unknown as Json,
  };
  fs.mkdirSync(RESULTS, { recursive: true });
  const out = path.join(RESULTS, `story-review-${pyStamp()}.json`);
  fs.writeFileSync(out, pyDumpsIndent(tagPythonFloats(report as unknown as Json, new Set(["dimensions"]), new Set(["verdict_accuracy", "taxonomy_accuracy", "gate_agreement"]))), "utf-8");
  const summary = Object.fromEntries(Object.entries(report).filter(([k]) => k !== "per_example"));
  console.log(pyDumpsIndent(tagPythonFloats(summary as unknown as Json, new Set(), new Set(["verdict_accuracy", "taxonomy_accuracy", "gate_agreement"]))));
  for (const p of perExample) {
    const mark = p.ok ? "ok " : "MISS";
    console.log(`  [${mark}] ${p.id}: expected=${p.expected} got=${p.got} kind=${p.failure_kind}`);
  }
  console.log(`report: ${out}`);
}

if (import.meta.url === `file://${process.argv[1]}`) await main();
